activerecord/lib/active_record/associations/preloader.rb

```diff
@@ -181,22 +174,18 @@ def preloaded_records
         end
 
         # Returns a class containing the logic needed to load preload the data
-        # and attach it to a relation. For example +Preloader::Association+ or
-        # +Preloader::HasManyThrough+. The class returned implements a `run` method
+        # and attach it to a relation: +Preloader::DirectAssociation+ or
+        # +Preloader::ThroughAssociation+. The class returned implements a `run` method
         # that accepts a preloader.
         def preloader_for(reflection, owners)
           if owners.first.association(reflection.name).loaded?
             return AlreadyLoaded
           end
           reflection.check_preloadable!
-
-          case reflection.macro
-          when :has_many
-            reflection.options[:through] ? HasManyThrough : HasMany
-          when :has_one
-            reflection.options[:through] ? HasOneThrough : HasOne
-          when :belongs_to
-            BelongsTo
+          if reflection.options[:through]
+            ThroughAssociation
+          else
+            DirectAssociation
```

Comment:
Moving most methods to new class will hurt the blameability.
Prefer reusing `Preloader::Association` to touch lesser lines.

Reply:
Yes, I was thinking about that too, but things change... The blame is still traceable manually via tracking particular line of code via `git log -p`. When I backtrace rails changes, it is almost always the case. The code moved from `Association` to `DirectAssociation` is easy to read and it didn't change for a really long period of time. So decided to sacrifice blame for ideal OOD.

Reply:
If it's moving mostly in unmodified hunks, I believe blame can follow it across files anyway.

(I haven't looked at the actual diff, so have no opinion on that beyond "+117 −195" sounding good.)

Reply:
At least we can remove preloader classes without touching existing code.
https://github.com/rails/rails/compare/master...kamipo:remove_useless_preloader_classes

Reply:
I'm more inclined to kamipo's refactoring.

Reply:
That is fine. You may accept that one.

Reply:
@kamipo mind to take care of it?

Reply:
Yeah, I will do that.